File: BinarySearchTree/BinarySearchTree.cs
namespace BST
{
    public class Node
    {
        public int Key;
        public Node Left;
        public Node Right;

        public Node(int item)
        {
            Key = item;
            Left = Right = default!;
        }
    }

    public class BinarySearchTree
    {
        public Node root; // just used for iterative version of insert
        public Node insert(Node node, int key)
        {
            if (node == null)
            {
                return new Node(key);
            }

            if (key < node.Key)
            {
                node.Left = insert(node.Left, key);
            }
            else
            {
                node.Right = insert(node.Right, key);
            }
            return node;
        }

        public void InsertIterative(int key)
        {
            Node node = new Node(key);
            if (root == null)
            {
                root = node;
                return;
            }
            Node prev = null!;
            Node temp = root;

            while (temp != null)
            {
                prev = temp;
                if (key < temp.Key)
                {

                    temp = temp.Left;
                }
                else
                {
                    temp = temp.Right;
                }
            }
            if (key < prev.Key)
            {
                prev.Left = node;
            }
            else
            {
                prev.Right = node;
            }
        }

        public Node Search(Node root, int key)
        {
            if (root == null || root.Key == key)
            {
                return root!;
            }
            if (key < root.Key)
            {
                return Search(root.Left, key);
            }
            return Search(root.Right, key);
        }
    }
}

File: Python/BinarySearchTree.py
# Binary Search Tree
class Node:
    def __init__(self, key) -> None:
        self.key = key
        self.left = None
        self.right = None


def insert(node, key):
    if node is None:
        return Node(key)
    
    if key < node.key:
        node.left = insert(node.left, key)
    elif key> node.key:
        node.right = insert(node.right, key)

    return node


def inorder(root):
    if root is not None:
        inorder(root.left)
        print(root.key, end=" ")
        inorder(root.right)


def height(node):
    if node is None:
        return 0
    else:
        lDepth = height(node.left)
        rDepth = height(node.right)

        return max(lDepth, rDepth) + 1


def printGivenLevel(root, level):
    if root is None:
        return
    
    if level == 1:
        print(root.key, end=" ")
    elif level > 1:
        printGivenLevel(root.left, level-1)
        printGivenLevel(root.right, level-1)
  

def printLevelOrder(root):
    h = height(root)
    for i in range(1, h+1):
        printGivenLevel(root, i)


def rightViewUtil(root, level, max_level):
    if root is None:
        return
    
    if max_level[0] < level:
        print("\t", root.key, end="")
        max_level[0] = level
    
    rightViewUtil(root.right, level+1, max_level)
    rightViewUtil(root.left, level+1, max_level)


def rightView(root):
    max_level = [0]
    rightViewUtil(root, 1, max_level)


def MinValueNode(node):
    current = node
    while current and current.left is not None:
        current = current.left
    
    return current

def deleteNode(root, key):
    if root is None:
        return root
    
    if key < root.key:
        root.left = deleteNode(root.left, key)
    
    elif key > root.key:
        root.right = deleteNode(root.right, key)
    
    else:
        if root.left is None:
            temp = root.right
            root = None
            return temp
        elif root.right is None:
            temp = root.left
            root = None
            return temp

        temp = MinValueNode(root.right)

        root.key = temp.key
        root.right = deleteNode(root.right, temp.key)

